File: problems/0005-longest-palindromic-substring/solution.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        /*
        Problem: 0005 - Longest Palindromic Substring.

        Approach:
        - Use dynamic programming to fill a table where dp[left][right] is true if s[left..right] is a palindrome.
        - Initialize all single-character and two-character palindromes.
        - Expand to longer substrings from the bottom up.

        Time Complexity: O(n^2), where n is the length of the string.
        Space Complexity: O(n^2), for the dp table.
        */

        int n = s.size();
        if (n == 0) return "";

        vector<vector<bool>> dp(n, vector<bool>(n, false));
        int max_length = 0;
        string res = "";

        // All substrings of length 1 are palindromes.
        for (int left = 0; left < n; ++left) {
            dp[left][left] = true;
            if (left != n - 1 && s[left] == s[left + 1]) {
                dp[left][left + 1] = true;
            }
        }

        // Check substrings of length ≥ 3.
        for (int left = n - 3; left >= 0; --left) {
            for (int right = left + 2; right < n; ++right) {
                if (s[left] == s[right] && dp[left + 1][right - 1]) {
                    dp[left][right] = true;
                }
            }
        }

        // Find the longest palindromic substring.
        for (int left = 0; left < n; ++left) {
            for (int right = left; right < n; ++right) {
                if (dp[left][right] && (right - left + 1 > max_length)) {
                    max_length = right - left + 1;
                    res = s.substr(left, max_length);
                }
            }
        }

        return res;
    }
};

File: problems/0005-longest-palindromic-substring/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        """
        Problem: 0005 - Longest Palindromic Substring.

        Approach:
        - Use dynamic programming to fill a table where dp[left][right] is True if s[left:right+1] is a palindrome.
        - Initialize all single-character and two-character substrings.
        - Expand to longer substrings from the bottom up.

        Time Complexity: O(n^2), where n is the length of the string.
        Space Complexity: O(n^2), for the dp table.
        """

        n = len(s)
        if n == 0:
            return ""

        # Initialize the DP table.
        dp = [[False for right in range(n)] for left in range(n)]

        # Mark all single-character substrings as palindromes.
        for left in range(n):
            dp[left][left] = True
            if left != n - 1:
                dp[left][left + 1] = (s[left] == s[left + 1])

        # Fill the table for substrings of length ≥ 3.
        for left in range(n - 3, -1, -1):
            for right in range(left + 2, n):
                dp[left][right] = dp[left + 1][right - 1] and s[left] == s[right]

        # Find the longest palindromic substring.
        max_length = 0
        res = ""
        for left in range(n):
            for right in range(left, n):
                if dp[left][right] and (right - left + 1) > max_length:
                    max_length = right - left + 1
                    res = s[left:right + 1]

        return res

